## Note to self:  turn this into a function or short script file to recreate quickly

In [2]:
import pandas as pd
import datetime as dt
Real_Prop = pd.read_csv("../data/EXTR_RPSale.csv")
Res_Bldg = pd.read_csv("../data/EXTR_ResBldg.csv")
Parcel = pd.read_csv("../data/EXTR_Parcel.csv", encoding='latin-1')
codes = pd.read_csv('../data/EXTR_LookUp.csv')

/opt/anaconda3/envs/oy-env/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3062: DtypeWarning: Columns (1,2) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
/opt/anaconda3/envs/oy-env/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3062: DtypeWarning: Columns (11) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


Since we are only interested in 2019 data, using datetime to convert documentdate and filtering for 2019 sales

In [3]:
Real_Prop['DocumentDate'] = pd.to_datetime(Real_Prop['DocumentDate'])
Real_Prop19 = Real_Prop[Real_Prop['DocumentDate'].dt.year == 2019]
Real_Prop19

ExciseTaxNbr   Major Minor DocumentDate  SalePrice    RecordingNbr  \
4             3027422  213043   120   2019-12-20     560000  20191226000848   
118           2999169  919715   200   2019-07-08     192000  20190712001080   
144           3000673  894444   200   2019-06-26     185000  20190722001395   
164           3002257  940652   630   2019-07-22     435000  20190730001339   
445           2980836  937630   695   2019-03-28     550000  20190404001008   
...               ...     ...   ...          ...        ...             ...   
2089056       3004408   66000  2210   2019-08-07   41040000  20190812000849   
2089057       3004408   66000  2225   2019-08-07   41040000  20190812000849   
2089058       3004408   66000  2195   2019-08-07   41040000  20190812000849   
2089059       3004408   66000  2220   2019-08-07   41040000  20190812000849   
2089082       3009227  192300   170   2019-09-03    1550000  20190909000838   

        Volume Page PlatNbr PlatType  ... PropertyType PrincipalUse  \
4                                     ...           11            6   
118                                   ...            3            2   
144                                   ...            3            2   
164                                   ...           11            6   
445                                   ...            3            6   
...        ...  ...     ...      ...  ...          ...          ...   
2089056                               ...            3            7   
2089057                               ...            3            7   
2089058                               ...            3            7   
2089059                               ...            3            7   
2089082                               ...           11            6   

        SaleInstrument AFForestLand  AFCurrentUseLand  AFNonProfitUse  \
4                    3            N                 N               N   
118                  3            N                 N               N   
144                  3            N                 N               N   
164                  3            N                 N               N   
445                  3            N                 N               N   
...                ...          ...               ...             ...   
2089056             22            N                 N               N   
2089057             22            N                 N               N   
2089058             22            N                 N               N   
2089059             22            N                 N               N   
2089082              3            N                 N               N   

         AFHistoricProperty SaleReason PropertyClass SaleWarning  
4                         N          1             8              
118                       N          1             3              
144                       N          1             3              
164                       N          1             8              
445                       N          1             8              
...                     ...        ...           ...         ...  
2089056                   N         18             2          45  
2089057                   N         18             2          45  
2089058                   N         18             2          45  
2089059                   N         18             2          45  
2089082                   N          1             8              

[61363 rows x 24 columns]

formatting column names to make them more Pythonic!

In [4]:
Real_Prop19 = Real_Prop19.rename(columns=str.lower)

In [5]:
Real_Prop19

excisetaxnbr   major minor documentdate  saleprice    recordingnbr  \
4             3027422  213043   120   2019-12-20     560000  20191226000848   
118           2999169  919715   200   2019-07-08     192000  20190712001080   
144           3000673  894444   200   2019-06-26     185000  20190722001395   
164           3002257  940652   630   2019-07-22     435000  20190730001339   
445           2980836  937630   695   2019-03-28     550000  20190404001008   
...               ...     ...   ...          ...        ...             ...   
2089056       3004408   66000  2210   2019-08-07   41040000  20190812000849   
2089057       3004408   66000  2225   2019-08-07   41040000  20190812000849   
2089058       3004408   66000  2195   2019-08-07   41040000  20190812000849   
2089059       3004408   66000  2220   2019-08-07   41040000  20190812000849   
2089082       3009227  192300   170   2019-09-03    1550000  20190909000838   

        volume page platnbr plattype  ... propertytype principaluse  \
4                                     ...           11            6   
118                                   ...            3            2   
144                                   ...            3            2   
164                                   ...           11            6   
445                                   ...            3            6   
...        ...  ...     ...      ...  ...          ...          ...   
2089056                               ...            3            7   
2089057                               ...            3            7   
2089058                               ...            3            7   
2089059                               ...            3            7   
2089082                               ...           11            6   

        saleinstrument afforestland  afcurrentuseland  afnonprofituse  \
4                    3            N                 N               N   
118                  3            N                 N               N   
144                  3            N                 N               N   
164                  3            N                 N               N   
445                  3            N                 N               N   
...                ...          ...               ...             ...   
2089056             22            N                 N               N   
2089057             22            N                 N               N   
2089058             22            N                 N               N   
2089059             22            N                 N               N   
2089082              3            N                 N               N   

         afhistoricproperty salereason propertyclass salewarning  
4                         N          1             8              
118                       N          1             3              
144                       N          1             3              
164                       N          1             8              
445                       N          1             8              
...                     ...        ...           ...         ...  
2089056                   N         18             2          45  
2089057                   N         18             2          45  
2089058                   N         18             2          45  
2089059                   N         18             2          45  
2089082                   N          1             8              

[61363 rows x 24 columns]

padding major and minor and adding them together after converting them to string type

In [6]:
Real_Prop19['major'] = Real_Prop19['major'].astype(str)
Real_Prop19['minor'] = Real_Prop19['minor'].astype(str)

In [7]:
def major_padded(row):
    if len(row) == 6:
        return row
    elif len(row) == 5:
        return '0'+row
    elif len(row) == 4:
        return '00'+row
    elif len(row) == 3:
        return '000'+row
    elif len(row) == 2:
        return '0000'+row
    elif len(row) == 1:
        return '00000'+row
    else:
        return '000000'

def minor_padded(row):
    if len(row) == 4:
        return row
    elif len(row) == 3: 
        return '0'+row
    elif len(row) == 2:
        return '00'+row
    elif len(row) == 1:
        return '000'+row
    else:
        return '0000'
    
Real_Prop19['major'] = Real_Prop19.major.apply(major_padded)
Real_Prop19['minor'] = Real_Prop19.minor.apply(minor_padded)
Real_Prop19['major_minor'] = Real_Prop19['major']+Real_Prop19['minor']

In [8]:
Real_Prop19

excisetaxnbr   major minor documentdate  saleprice    recordingnbr  \
4             3027422  213043  0120   2019-12-20     560000  20191226000848   
118           2999169  919715  0200   2019-07-08     192000  20190712001080   
144           3000673  894444  0200   2019-06-26     185000  20190722001395   
164           3002257  940652  0630   2019-07-22     435000  20190730001339   
445           2980836  937630  0695   2019-03-28     550000  20190404001008   
...               ...     ...   ...          ...        ...             ...   
2089056       3004408  066000  2210   2019-08-07   41040000  20190812000849   
2089057       3004408  066000  2225   2019-08-07   41040000  20190812000849   
2089058       3004408  066000  2195   2019-08-07   41040000  20190812000849   
2089059       3004408  066000  2220   2019-08-07   41040000  20190812000849   
2089082       3009227  192300  0170   2019-09-03    1550000  20190909000838   

        volume page platnbr plattype  ... principaluse saleinstrument  \
4                                     ...            6              3   
118                                   ...            2              3   
144                                   ...            2              3   
164                                   ...            6              3   
445                                   ...            6              3   
...        ...  ...     ...      ...  ...          ...            ...   
2089056                               ...            7             22   
2089057                               ...            7             22   
2089058                               ...            7             22   
2089059                               ...            7             22   
2089082                               ...            6              3   

        afforestland afcurrentuseland  afnonprofituse  afhistoricproperty  \
4                  N                N               N                   N   
118                N                N               N                   N   
144                N                N               N                   N   
164                N                N               N                   N   
445                N                N               N                   N   
...              ...              ...             ...                 ...   
2089056            N                N               N                   N   
2089057            N                N               N                   N   
2089058            N                N               N                   N   
2089059            N                N               N                   N   
2089082            N                N               N                   N   

         salereason propertyclass salewarning major_minor  
4                 1             8              2130430120  
118               1             3              9197150200  
144               1             3              8944440200  
164               1             8              9406520630  
445               1             8              9376300695  
...             ...           ...         ...         ...  
2089056          18             2          45  0660002210  
2089057          18             2          45  0660002225  
2089058          18             2          45  0660002195  
2089059          18             2          45  0660002220  
2089082           1             8              1923000170  

[61363 rows x 25 columns]

moving the major_minor column to column1 position

In [9]:
col_name = 'major_minor'
last_col = Real_Prop19.pop(col_name)
Real_Prop19.insert(1, col_name, last_col)

In [10]:
Real_Prop19

excisetaxnbr major_minor   major minor documentdate  saleprice  \
4             3027422  2130430120  213043  0120   2019-12-20     560000   
118           2999169  9197150200  919715  0200   2019-07-08     192000   
144           3000673  8944440200  894444  0200   2019-06-26     185000   
164           3002257  9406520630  940652  0630   2019-07-22     435000   
445           2980836  9376300695  937630  0695   2019-03-28     550000   
...               ...         ...     ...   ...          ...        ...   
2089056       3004408  0660002210  066000  2210   2019-08-07   41040000   
2089057       3004408  0660002225  066000  2225   2019-08-07   41040000   
2089058       3004408  0660002195  066000  2195   2019-08-07   41040000   
2089059       3004408  0660002220  066000  2220   2019-08-07   41040000   
2089082       3009227  1923000170  192300  0170   2019-09-03    1550000   

           recordingnbr volume page platnbr  ... propertytype principaluse  \
4        20191226000848                      ...           11            6   
118      20190712001080                      ...            3            2   
144      20190722001395                      ...            3            2   
164      20190730001339                      ...           11            6   
445      20190404001008                      ...            3            6   
...                 ...    ...  ...     ...  ...          ...          ...   
2089056  20190812000849                      ...            3            7   
2089057  20190812000849                      ...            3            7   
2089058  20190812000849                      ...            3            7   
2089059  20190812000849                      ...            3            7   
2089082  20190909000838                      ...           11            6   

        saleinstrument afforestland afcurrentuseland  afnonprofituse  \
4                    3            N                N               N   
118                  3            N                N               N   
144                  3            N                N               N   
164                  3            N                N               N   
445                  3            N                N               N   
...                ...          ...              ...             ...   
2089056             22            N                N               N   
2089057             22            N                N               N   
2089058             22            N                N               N   
2089059             22            N                N               N   
2089082              3            N                N               N   

         afhistoricproperty  salereason propertyclass salewarning  
4                         N           1             8              
118                       N           1             3              
144                       N           1             3              
164                       N           1             8              
445                       N           1             8              
...                     ...         ...           ...         ...  
2089056                   N          18             2          45  
2089057                   N          18             2          45  
2089058                   N          18             2          45  
2089059                   N          18             2          45  
2089082                   N           1             8              

[61363 rows x 25 columns]

investigating propertytype to see the unique entries

In [11]:
Real_Prop19.propertytype.unique()

array([11,  3, 14, 51,  1, 10,  2, 59, 91, 16,  4, 12,  5,  0, 13, 58, 69,
       50, 99, 63, 18, 45, 79, 83, 65, 94,  6, 53, 68, 44, 54, 39, 85, 64,
       86, 55, 47, 19, 88, 17, 33, 15, 96, 27, 23, 46, 48, 52, 61, 81,  9,
       66, 21, 74])

The only propertytypes we are interested in are:

11 - Household, single family units

12 - Multiple family residence (Residential, 2-4 units)

14 - Residential condominiums

18 - All other residential not elsewhere coded

19 - Vacation and cabin (decided to not include vacation homes, probably first time homebuyers don't really go for these)

In [12]:
Real_Prop19 = Real_Prop19[Real_Prop19['propertytype'].isin([11, 12, 14, 18])]
Real_Prop19

excisetaxnbr major_minor   major minor documentdate  saleprice  \
4             3027422  2130430120  213043  0120   2019-12-20     560000   
164           3002257  9406520630  940652  0630   2019-07-22     435000   
456           2970575  8669201280  866920  1280   2019-01-16          0   
497           3025919  0565250340  056525  0340   2019-12-12     311000   
517           2975376  0984000450  098400  0450   2019-02-20     409950   
...               ...         ...     ...   ...          ...        ...   
2088838       2986110  8732340080  873234  0080   2019-04-10    2375000   
2088923       3012006  3904970150  390497  0150   2019-09-18     850000   
2088953       2974971  2460900090  246090  0090   2019-02-13     686000   
2088987       2998779  1438000420  143800  0420   2019-07-09     695000   
2089082       3009227  1923000170  192300  0170   2019-09-03    1550000   

           recordingnbr volume page platnbr  ... propertytype principaluse  \
4        20191226000848                      ...           11            6   
164      20190730001339                      ...           11            6   
456      20190118000160                      ...           14            2   
497      20191217001069                      ...           14            2   
517      20190228000242                      ...           11            6   
...                 ...    ...  ...     ...  ...          ...          ...   
2088838  20190503001267                      ...           14            2   
2088923  20190925001198                      ...           11            6   
2088953  20190226000416                      ...           11            6   
2088987  20190711000444                      ...           11            6   
2089082  20190909000838                      ...           11            6   

        saleinstrument afforestland afcurrentuseland  afnonprofituse  \
4                    3            N                N               N   
164                  3            N                N               N   
456                  3            N                N               N   
497                  3            N                N               N   
517                  3            N                N               N   
...                ...          ...              ...             ...   
2088838              3            N                N               N   
2088923              3            N                N               N   
2088953              3            N                N               N   
2088987              3            N                N               N   
2089082              3            N                N               N   

         afhistoricproperty  salereason propertyclass salewarning  
4                         N           1             8              
164                       N           1             8              
456                       N           1             3              
497                       N           1             3              
517                       N           1             8              
...                     ...         ...           ...         ...  
2088838                   N           1             3              
2088923                   N           1             8              
2088953                   N           1             8              
2088987                   N           1             8              
2089082                   N           1             8              

[30516 rows x 25 columns]

This has shrunk our data from 61363 entries to 30516

Let's only keep entries that have a Saleprice of more than 0

In [13]:
Real_Prop19 = Real_Prop19[Real_Prop19['saleprice'] > 0]
Real_Prop19

excisetaxnbr major_minor   major minor documentdate  saleprice  \
4             3027422  2130430120  213043  0120   2019-12-20     560000   
164           3002257  9406520630  940652  0630   2019-07-22     435000   
497           3025919  0565250340  056525  0340   2019-12-12     311000   
517           2975376  0984000450  098400  0450   2019-02-20     409950   
628           2976405  0302000645  030200  0645   2019-03-04     380000   
...               ...         ...     ...   ...          ...        ...   
2088838       2986110  8732340080  873234  0080   2019-04-10    2375000   
2088923       3012006  3904970150  390497  0150   2019-09-18     850000   
2088953       2974971  2460900090  246090  0090   2019-02-13     686000   
2088987       2998779  1438000420  143800  0420   2019-07-09     695000   
2089082       3009227  1923000170  192300  0170   2019-09-03    1550000   

           recordingnbr volume page platnbr  ... propertytype principaluse  \
4        20191226000848                      ...           11            6   
164      20190730001339                      ...           11            6   
497      20191217001069                      ...           14            2   
517      20190228000242                      ...           11            6   
628      20190306000786                      ...           11            6   
...                 ...    ...  ...     ...  ...          ...          ...   
2088838  20190503001267                      ...           14            2   
2088923  20190925001198                      ...           11            6   
2088953  20190226000416                      ...           11            6   
2088987  20190711000444                      ...           11            6   
2089082  20190909000838                      ...           11            6   

        saleinstrument afforestland afcurrentuseland  afnonprofituse  \
4                    3            N                N               N   
164                  3            N                N               N   
497                  3            N                N               N   
517                  3            N                N               N   
628                  3            N                N               N   
...                ...          ...              ...             ...   
2088838              3            N                N               N   
2088923              3            N                N               N   
2088953              3            N                N               N   
2088987              3            N                N               N   
2089082              3            N                N               N   

         afhistoricproperty  salereason propertyclass salewarning  
4                         N           1             8              
164                       N           1             8              
497                       N           1             3              
517                       N           1             8              
628                       N           1             8          10  
...                     ...         ...           ...         ...  
2088838                   N           1             3              
2088923                   N           1             8              
2088953                   N           1             8              
2088987                   N           1             8              
2089082                   N           1             8              

[21372 rows x 25 columns]

In [13]:
Real_Prop19.saleprice.agg(['min', 'max', 'std', 'mean'])

min     1.000000e+01
max     5.300000e+07
std     6.747552e+05
mean    7.585140e+05
Name: saleprice, dtype: float64

we now have 21372 entries, checking to see the principal use case for these entries 

In [14]:
Real_Prop19['principaluse'].unique()

array([6, 2, 7])

6 - Residential

2 - Condominium

7 - Commercial

Lets remove entries relating to commercial use

In [14]:
Real_Prop19 = Real_Prop19[Real_Prop19['principaluse'].isin([6,2])]
Real_Prop19

excisetaxnbr major_minor   major minor documentdate  saleprice  \
4             3027422  2130430120  213043  0120   2019-12-20     560000   
164           3002257  9406520630  940652  0630   2019-07-22     435000   
497           3025919  0565250340  056525  0340   2019-12-12     311000   
517           2975376  0984000450  098400  0450   2019-02-20     409950   
628           2976405  0302000645  030200  0645   2019-03-04     380000   
...               ...         ...     ...   ...          ...        ...   
2088838       2986110  8732340080  873234  0080   2019-04-10    2375000   
2088923       3012006  3904970150  390497  0150   2019-09-18     850000   
2088953       2974971  2460900090  246090  0090   2019-02-13     686000   
2088987       2998779  1438000420  143800  0420   2019-07-09     695000   
2089082       3009227  1923000170  192300  0170   2019-09-03    1550000   

           recordingnbr volume page platnbr  ... propertytype principaluse  \
4        20191226000848                      ...           11            6   
164      20190730001339                      ...           11            6   
497      20191217001069                      ...           14            2   
517      20190228000242                      ...           11            6   
628      20190306000786                      ...           11            6   
...                 ...    ...  ...     ...  ...          ...          ...   
2088838  20190503001267                      ...           14            2   
2088923  20190925001198                      ...           11            6   
2088953  20190226000416                      ...           11            6   
2088987  20190711000444                      ...           11            6   
2089082  20190909000838                      ...           11            6   

        saleinstrument afforestland afcurrentuseland  afnonprofituse  \
4                    3            N                N               N   
164                  3            N                N               N   
497                  3            N                N               N   
517                  3            N                N               N   
628                  3            N                N               N   
...                ...          ...              ...             ...   
2088838              3            N                N               N   
2088923              3            N                N               N   
2088953              3            N                N               N   
2088987              3            N                N               N   
2089082              3            N                N               N   

         afhistoricproperty  salereason propertyclass salewarning  
4                         N           1             8              
164                       N           1             8              
497                       N           1             3              
517                       N           1             8              
628                       N           1             8          10  
...                     ...         ...           ...         ...  
2088838                   N           1             3              
2088923                   N           1             8              
2088953                   N           1             8              
2088987                   N           1             8              
2089082                   N           1             8              

[21260 rows x 25 columns]

down to 21260 entries. lets give one last check for property class

In [16]:
Real_Prop19['propertyclass'].unique()

array([8, 3, 7, 9])

7 - Res-Land only, we can remove that from our dataframe to only reflect residential property

keeping only:

8 - Residential- Improved Property

3 - Condominium

9 - Res/Condo mobile

In [15]:
Real_Prop19 = Real_Prop19[Real_Prop19['propertyclass'].isin([8,3,9])]
Real_Prop19

excisetaxnbr major_minor   major minor documentdate  saleprice  \
4             3027422  2130430120  213043  0120   2019-12-20     560000   
164           3002257  9406520630  940652  0630   2019-07-22     435000   
497           3025919  0565250340  056525  0340   2019-12-12     311000   
517           2975376  0984000450  098400  0450   2019-02-20     409950   
628           2976405  0302000645  030200  0645   2019-03-04     380000   
...               ...         ...     ...   ...          ...        ...   
2088838       2986110  8732340080  873234  0080   2019-04-10    2375000   
2088923       3012006  3904970150  390497  0150   2019-09-18     850000   
2088953       2974971  2460900090  246090  0090   2019-02-13     686000   
2088987       2998779  1438000420  143800  0420   2019-07-09     695000   
2089082       3009227  1923000170  192300  0170   2019-09-03    1550000   

           recordingnbr volume page platnbr  ... propertytype principaluse  \
4        20191226000848                      ...           11            6   
164      20190730001339                      ...           11            6   
497      20191217001069                      ...           14            2   
517      20190228000242                      ...           11            6   
628      20190306000786                      ...           11            6   
...                 ...    ...  ...     ...  ...          ...          ...   
2088838  20190503001267                      ...           14            2   
2088923  20190925001198                      ...           11            6   
2088953  20190226000416                      ...           11            6   
2088987  20190711000444                      ...           11            6   
2089082  20190909000838                      ...           11            6   

        saleinstrument afforestland afcurrentuseland  afnonprofituse  \
4                    3            N                N               N   
164                  3            N                N               N   
497                  3            N                N               N   
517                  3            N                N               N   
628                  3            N                N               N   
...                ...          ...              ...             ...   
2088838              3            N                N               N   
2088923              3            N                N               N   
2088953              3            N                N               N   
2088987              3            N                N               N   
2089082              3            N                N               N   

         afhistoricproperty  salereason propertyclass salewarning  
4                         N           1             8              
164                       N           1             8              
497                       N           1             3              
517                       N           1             8              
628                       N           1             8          10  
...                     ...         ...           ...         ...  
2088838                   N           1             3              
2088923                   N           1             8              
2088953                   N           1             8              
2088987                   N           1             8              
2089082                   N           1             8              

[20999 rows x 25 columns]

investigated salewarnings and decided to skip and power through the rest.....

In [18]:
Real_Prop19.salewarning.value_counts()

            19598
15            281
26            214
40            184
41            152
            ...  
57              1
38              1
45              1
12 46 51        1
44              1
Name: salewarning, Length: 98, dtype: int64

Now checking to see if major_minor are unique or not

In [19]:
len(Real_Prop19) - len(Real_Prop19.major_minor.unique())

377

moving along and not worrying about the major_minor repeats, since it could be due to successive sale of the same property in 2019 or other reasons... can explore later.....

Lets drop the blank columns 'volume', 'page', 'platnbr', 'plattype', 'platlot', 'platblock', 'sellername', 'buyername'


In [16]:
cols_to_drop = ['volume', 'page', 'platnbr', 'plattype', 'platlot', 'platblock', 'sellername', 'buyername']
Real_Prop19.drop(columns=cols_to_drop, inplace=True)

/opt/anaconda3/envs/oy-env/lib/python3.8/site-packages/pandas/core/frame.py:3990: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [21]:
Real_Prop19

excisetaxnbr major_minor   major minor documentdate  saleprice  \
4             3027422  2130430120  213043  0120   2019-12-20     560000   
164           3002257  9406520630  940652  0630   2019-07-22     435000   
497           3025919  0565250340  056525  0340   2019-12-12     311000   
517           2975376  0984000450  098400  0450   2019-02-20     409950   
628           2976405  0302000645  030200  0645   2019-03-04     380000   
...               ...         ...     ...   ...          ...        ...   
2088838       2986110  8732340080  873234  0080   2019-04-10    2375000   
2088923       3012006  3904970150  390497  0150   2019-09-18     850000   
2088953       2974971  2460900090  246090  0090   2019-02-13     686000   
2088987       2998779  1438000420  143800  0420   2019-07-09     695000   
2089082       3009227  1923000170  192300  0170   2019-09-03    1550000   

           recordingnbr  propertytype  principaluse  saleinstrument  \
4        20191226000848            11             6               3   
164      20190730001339            11             6               3   
497      20191217001069            14             2               3   
517      20190228000242            11             6               3   
628      20190306000786            11             6               3   
...                 ...           ...           ...             ...   
2088838  20190503001267            14             2               3   
2088923  20190925001198            11             6               3   
2088953  20190226000416            11             6               3   
2088987  20190711000444            11             6               3   
2089082  20190909000838            11             6               3   

        afforestland afcurrentuseland afnonprofituse afhistoricproperty  \
4                  N                N              N                  N   
164                N                N              N                  N   
497                N                N              N                  N   
517                N                N              N                  N   
628                N                N              N                  N   
...              ...              ...            ...                ...   
2088838            N                N              N                  N   
2088923            N                N              N                  N   
2088953            N                N              N                  N   
2088987            N                N              N                  N   
2089082            N                N              N                  N   

         salereason  propertyclass salewarning  
4                 1              8              
164               1              8              
497               1              3              
517               1              8              
628               1              8          10  
...             ...            ...         ...  
2088838           1              3              
2088923           1              8              
2088953           1              8              
2088987           1              8              
2089082           1              8              

[20999 rows x 17 columns]

At present we have 20999 entries that satisfy all our conditions for residential propert sales in King county in 2019

Now bringing in the Residential Buildings table to mearge with our current datframe for features of the properties we have narrowed down.

We'll first clean up column names, pad up major and minor and combine them.

In [17]:
Res_Bldg.rename(columns=str.lower, inplace=True)

In [18]:
Res_Bldg['major'] = Res_Bldg['major'].astype(str)
Res_Bldg['minor'] = Res_Bldg['minor'].astype(str)

In [19]:
Res_Bldg['major'] = Res_Bldg['major'].apply(major_padded)
Res_Bldg['minor'] = Res_Bldg['minor'].apply(minor_padded)

In [20]:
Res_Bldg['major_minor'] = Res_Bldg['major']+Res_Bldg['minor']

In [21]:
col_name = 'major_minor'
last_col = Res_Bldg.pop(col_name)
Res_Bldg.insert(0, col_name, last_col)

In [27]:
Res_Bldg.head(2)

major_minor   major minor  bldgnbr  nbrlivingunits  \
0  0461000935  046100  0935        1               1   
1  0461001165  046100  1165        1               1   

                     address buildingnumber fraction directionprefix  \
0  7349   10TH AVE NW  98117          7349                             
1  7342   11TH AVE NW  98117          7342                             

                  streetname  ... fpmultistory fpfreestanding fpadditional  \
0  10TH                       ...            0              0            0   
1  11TH                       ...            0              0            0   

   yrbuilt  yrrenovated  pcntcomplete  obsolescence  pcntnetcondition  \
0     1927            0             0             0                 0   
1     2019            0            53             0                 0   

   condition  addnlcost  
0          3          0  
1          3          0  

[2 rows x 51 columns]

In [28]:
Res_Bldg.loc[1]

major_minor                          0461001165
major                                    046100
minor                                      1165
bldgnbr                                       1
nbrlivingunits                                1
address               7342   11TH AVE NW  98117
buildingnumber                            7342 
fraction                                       
directionprefix                                
streetname            11TH                     
streettype                               AVE   
directionsuffix                              NW
zipcode                                   98117
stories                                       2
bldggrade                                     8
bldggradevar                                  0
sqft1stfloor                               1600
sqfthalffloor                                 0
sqft2ndfloor                               1560
sqftupperfloor                                0
sqftunfinfull                           

In [29]:
Res_Bldg.columns

Index(['major_minor', 'major', 'minor', 'bldgnbr', 'nbrlivingunits', 'address',
       'buildingnumber', 'fraction', 'directionprefix', 'streetname',
       'streettype', 'directionsuffix', 'zipcode', 'stories', 'bldggrade',
       'bldggradevar', 'sqft1stfloor', 'sqfthalffloor', 'sqft2ndfloor',
       'sqftupperfloor', 'sqftunfinfull', 'sqftunfinhalf', 'sqfttotliving',
       'sqfttotbasement', 'sqftfinbasement', 'finbasementgrade',
       'sqftgaragebasement', 'sqftgarageattached', 'daylightbasement',
       'sqftopenporch', 'sqftenclosedporch', 'sqftdeck', 'heatsystem',
       'heatsource', 'brickstone', 'viewutilization', 'bedrooms',
       'bathhalfcount', 'bath3qtrcount', 'bathfullcount', 'fpsinglestory',
       'fpmultistory', 'fpfreestanding', 'fpadditional', 'yrbuilt',
       'yrrenovated', 'pcntcomplete', 'obsolescence', 'pcntnetcondition',
       'condition', 'addnlcost'],
      dtype='object')

merging together Property sales in 2019 `Real_Prop19` with `Res_Bldg` purely based on their `major_minor`

In [22]:
prop_res = Real_Prop19.merge(Res_Bldg, on='major_minor', suffixes=("", "_resbldg"))

In [31]:
prop_res.describe()

excisetaxnbr     saleprice  propertytype  principaluse  saleinstrument  \
count  1.877100e+04  1.877100e+04  18771.000000       18771.0    18771.000000   
mean   2.999465e+06  7.955853e+05     11.028075           6.0        3.085185   
std    1.666185e+04  5.905325e+05      0.313453           0.0        1.261554   
min    2.968844e+06  1.000000e+01     11.000000           6.0        2.000000   
25%    2.984984e+06  4.700000e+05     11.000000           6.0        3.000000   
50%    3.000260e+06  6.564000e+05     11.000000           6.0        3.000000   
75%    3.012852e+06  9.067500e+05     11.000000           6.0        3.000000   
max    3.050564e+06  1.605000e+07     18.000000           6.0       26.000000   

         salereason  propertyclass       bldgnbr  nbrlivingunits  \
count  18771.000000        18771.0  18771.000000    18771.000000   
mean       1.004528            8.0      1.007671        1.014597   
std        0.277424            0.0      0.100848        0.139258   
min        1.000000            8.0      1.000000        1.000000   
25%        1.000000            8.0      1.000000        1.000000   
50%        1.000000            8.0      1.000000        1.000000   
75%        1.000000            8.0      1.000000        1.000000   
max       18.000000            8.0      7.000000        3.000000   

            stories  ...  fpmultistory  fpfreestanding  fpadditional  \
count  18771.000000  ...  18771.000000    18771.000000  18771.000000   
mean       1.517127  ...      0.302488        0.067871      0.174684   
std        0.534168  ...      0.500957        0.260275      0.398065   
min        1.000000  ...      0.000000        0.000000      0.000000   
25%        1.000000  ...      0.000000        0.000000      0.000000   
50%        1.500000  ...      0.000000        0.000000      0.000000   
75%        2.000000  ...      1.000000        0.000000      0.000000   
max        3.500000  ...      5.000000        2.000000      3.000000   

            yrbuilt   yrrenovated  pcntcomplete  obsolescence  \
count  18771.000000  18771.000000  18771.000000  18771.000000   
mean    1974.720633    100.592297      0.270630      0.101113   
std       30.384858    437.165864      4.680644      2.202412   
min     1900.000000      0.000000      0.000000      0.000000   
25%     1954.000000      0.000000      0.000000      0.000000   
50%     1978.000000      0.000000      0.000000      0.000000   
75%     2001.000000      0.000000      0.000000      0.000000   
max     2020.000000   2020.000000    100.000000     99.000000   

       pcntnetcondition     condition     addnlcost  
count      18771.000000  18771.000000  18771.000000  
mean           0.004795      3.463428    715.271802  
std            0.374718      0.694128   2565.710429  
min            0.000000      1.000000      0.000000  
25%            0.000000      3.000000      0.000000  
50%            0.000000      3.000000      0.000000  
75%            0.000000      4.000000      0.000000  
max           40.000000      5.000000  60000.000000  

[8 rows x 45 columns]

Down to 18771 entries, this could be due to having entries in the Real Property sales dataframe that were still reflecting property that were not purely residential. Merging with Residential Buildings could have narrowed this down further for us.... 

bringing in parcel table and cleaning it up

In [23]:
Parcel.rename(columns=str.lower, inplace=True)
Parcel['major'] = Parcel['major'].astype(str)
Parcel['minor'] = Parcel['minor'].astype(str)
Parcel['major'] = Parcel['major'].apply(major_padded)
Parcel['minor'] = Parcel['minor'].apply(minor_padded)
Parcel['major_minor'] = Parcel['major']+Parcel['minor']
last_col = Parcel.pop(col_name)
Parcel.insert(0, col_name, last_col)

In [33]:
Parcel.columns

Index(['major_minor', 'major', 'minor', 'propname', 'platname', 'platlot',
       'platblock', 'range', 'township', 'section', 'quartersection',
       'proptype', 'area', 'subarea', 'specarea', 'specsubarea',
       'districtname', 'levycode', 'currentzoning', 'hbuasifvacant',
       'hbuasimproved', 'presentuse', 'sqftlot', 'watersystem', 'sewersystem',
       'access', 'topography', 'streetsurface', 'restrictiveszshape',
       'inadequateparking', 'pcntunusable', 'unbuildable', 'mtrainier',
       'olympics', 'cascades', 'territorial', 'seattleskyline', 'pugetsound',
       'lakewashington', 'lakesammamish', 'smalllakerivercreek', 'otherview',
       'wfntlocation', 'wfntfootage', 'wfntbank', 'wfntpoorquality',
       'wfntrestrictedaccess', 'wfntaccessrights', 'wfntproximityinfluence',
       'tidelandshoreland', 'lotdepthfactor', 'trafficnoise', 'airportnoise',
       'powerlines', 'othernuisances', 'nbrbldgsites', 'contamination',
       'dnrlease', 'adjacentgolffairway', 'adjac

In [34]:
Parcel.loc[1]

major_minor           1326069228
major                     132606
minor                       9228
propname                        
platname                     NaN
                         ...    
speciesofconcern               N
sensitiveareatract             N
waterproblems                  N
transpconcurrency              N
otherproblems                  N
Name: 1, Length: 82, dtype: object

In [24]:
prop_res = prop_res.merge(Parcel, on='major_minor', suffixes=("", "_parcel"))

In [36]:
prop_res.head()

excisetaxnbr major_minor   major minor documentdate  saleprice  \
0       3027422  2130430120  213043  0120   2019-12-20     560000   
1       3002257  9406520630  940652  0630   2019-07-22     435000   
2       2975376  0984000450  098400  0450   2019-02-20     409950   
3       2976405  0302000645  030200  0645   2019-03-04     380000   
4       3012934  6154900010  615490  0010   2019-09-24     775000   

     recordingnbr  propertytype  principaluse  saleinstrument  ...  \
0  20191226000848            11             6               3  ...   
1  20190730001339            11             6               3  ...   
2  20190228000242            11             6               3  ...   
3  20190306000786            11             6               3  ...   
4  20190930002488            11             6               3  ...   

  seismichazard landslidehazard steepslopehazard stream  wetland  \
0             N               N                N      N        N   
1             N               N                N      N        N   
2             N               N                N      N        N   
3             N               N                N      N        N   
4             N               N                N      N        N   

   speciesofconcern sensitiveareatract waterproblems transpconcurrency  \
0                 N                  N             N                 N   
1                 N                  N             N                 N   
2                 N                  N             N                 N   
3                 N                  N             N                 N   
4                 N                  N             N                 N   

   otherproblems  
0              N  
1              N  
2              N  
3              N  
4              N  

[5 rows x 148 columns]

In [37]:
prop_res.shape

(18771, 148)

no loss of entries here, still at 18771!

In [38]:
prop_res['proptype'].unique()

array(['R'], dtype=object)

R - Residential, propert type included in this database only reflects residential property. That's a good sign!

In [39]:
prop_res.describe()

excisetaxnbr     saleprice  propertytype  principaluse  saleinstrument  \
count  1.877100e+04  1.877100e+04  18771.000000       18771.0    18771.000000   
mean   2.999465e+06  7.955853e+05     11.028075           6.0        3.085185   
std    1.666185e+04  5.905325e+05      0.313453           0.0        1.261554   
min    2.968844e+06  1.000000e+01     11.000000           6.0        2.000000   
25%    2.984984e+06  4.700000e+05     11.000000           6.0        3.000000   
50%    3.000260e+06  6.564000e+05     11.000000           6.0        3.000000   
75%    3.012852e+06  9.067500e+05     11.000000           6.0        3.000000   
max    3.050564e+06  1.605000e+07     18.000000           6.0       26.000000   

         salereason  propertyclass       bldgnbr  nbrlivingunits  \
count  18771.000000        18771.0  18771.000000    18771.000000   
mean       1.004528            8.0      1.007671        1.014597   
std        0.277424            0.0      0.100848        0.139258   
min        1.000000            8.0      1.000000        1.000000   
25%        1.000000            8.0      1.000000        1.000000   
50%        1.000000            8.0      1.000000        1.000000   
75%        1.000000            8.0      1.000000        1.000000   
max       18.000000            8.0      7.000000        3.000000   

            stories  ...  wfntpoorquality  wfntrestrictedaccess  \
count  18771.000000  ...     18771.000000          18771.000000   
mean       1.517127  ...         0.000533              0.005434   
std        0.534168  ...         0.023076              0.107632   
min        1.000000  ...         0.000000              0.000000   
25%        1.000000  ...         0.000000              0.000000   
50%        1.500000  ...         0.000000              0.000000   
75%        2.000000  ...         0.000000              0.000000   
max        3.500000  ...         1.000000              3.000000   

       tidelandshoreland  lotdepthfactor  trafficnoise  airportnoise  \
count       18771.000000    18771.000000  18771.000000       18771.0   
mean            0.015130        0.052688      0.200202           0.0   
std             0.172373        2.094350      0.545121           0.0   
min             0.000000        0.000000      0.000000           0.0   
25%             0.000000        0.000000      0.000000           0.0   
50%             0.000000        0.000000      0.000000           0.0   
75%             0.000000        0.000000      0.000000           0.0   
max             3.000000      105.000000      3.000000           0.0   

       nbrbldgsites  contamination  historicsite  currentusedesignation  
count  18771.000000        18771.0  18771.000000           18771.000000  
mean       0.005967            0.0      0.000906               0.003143  
std        0.107108            0.0      0.051086               0.095953  
min        0.000000            0.0      0.000000               0.000000  
25%        0.000000            0.0      0.000000               0.000000  
50%        0.000000            0.0      0.000000               0.000000  
75%        0.000000            0.0      0.000000               0.000000  
max        5.000000            0.0      3.000000               3.000000  

[8 rows x 88 columns]

total count 18771

choosing all the features of a residential property that would have a bearing on its saleprice:

**Residential building features:** (Specific to the unit)

total living square feet, year built, year renovated, bedrooms, bath(half, quarter, full), porch square feet(open, enclosed), deck, total bsement square feet, finished basement grade, daylight basement, heat system, heat source, condition

**Parcel features:** (specifi to the location of the lot)

HBU as is vacant, parking, transportation concurrency, views of Mt.Rainier, Olympics, Cascades, Territorial, Seattle skyline, Puget sound, Lake washington, Lake sammamish, small lake/river/creek, waterfront location, waterfront access rights trffic noise, airport noise, power lines, other nuisances, adjacent greenbelt.  (may want to include hazard metrics) 

In [25]:
df = prop_res.loc[:, ['saleprice', 'sqfttotliving', 'yrbuilt', 'yrrenovated', 'bedrooms',
                      'zipcode', 'sqfttotbasement', 'sqftfinbasement', 'sqftopenporch', 
                      'sqftenclosedporch', 'sqftdeck','heatsystem', 'heatsource', 'bathhalfcount', 
                      'bath3qtrcount', 'bathfullcount','condition', 'viewutilization', 'sqftgarageattached',
                      'daylightbasement', 'finbasementgrade', 'hbuasifvacant', 'inadequateparking',
                      'mtrainier', 'olympics', 'cascades', 'territorial', 'seattleskyline', 'pugetsound',
                      'lakewashington', 'lakesammamish', 'smalllakerivercreek', 'otherview', 'wfntlocation',
                      'trafficnoise', 'airportnoise', 'powerlines', 'othernuisances', 'adjacentgreenbelt']]

In [26]:
df.head()

saleprice  sqfttotliving  yrbuilt  yrrenovated  bedrooms zipcode  \
0     560000           1930     1989            0         3   98019   
1     435000           1975     1994            0         3   98038   
2     409950           1850     1968            0         3   98058   
3     380000           1750     1963            0         4   98001   
4     775000           2080     1947            0         4   98177   

   sqfttotbasement  sqftfinbasement  sqftopenporch  sqftenclosedporch  ...  \
0                0                0            180                  0  ...   
1                0                0             45                  0  ...   
2             1220              600              0                  0  ...   
3              920              460            670                  0  ...   
4             1040             1040              0                  0  ...   

   lakewashington  lakesammamish  smalllakerivercreek  otherview  \
0               0              0                    0          0   
1               0              0                    0          0   
2               0              0                    0          0   
3               0              0                    0          0   
4               0              0                    0          0   

   wfntlocation  trafficnoise  airportnoise powerlines  othernuisances  \
0             0             0             0          N               N   
1             0             0             0          N               N   
2             0             0             0          N               N   
3             0             0             0          N               N   
4             0             0             0          N               N   

  adjacentgreenbelt  
0                 N  
1                 N  
2                 N  
3                 N  
4                 N  

[5 rows x 39 columns]

In [27]:
df.describe()

saleprice  sqfttotliving       yrbuilt   yrrenovated      bedrooms  \
count  1.877100e+04   18771.000000  18771.000000  18771.000000  18771.000000   
mean   7.955853e+05    2145.487774   1974.720633    100.592297      3.427894   
std    5.905325e+05     968.251055     30.384858    437.165864      0.930544   
min    1.000000e+01     250.000000   1900.000000      0.000000      0.000000   
25%    4.700000e+05    1450.000000   1954.000000      0.000000      3.000000   
50%    6.564000e+05    1970.000000   1978.000000      0.000000      3.000000   
75%    9.067500e+05    2630.000000   2001.000000      0.000000      4.000000   
max    1.605000e+07   10910.000000   2020.000000   2020.000000     13.000000   

       sqfttotbasement  sqftfinbasement  sqftopenporch  sqftenclosedporch  \
count     18771.000000     18771.000000   18771.000000       18771.000000   
mean        481.943264       303.638059      72.835384           5.421501   
std         581.411684       453.159932     149.736725          36.765319   
min           0.000000         0.000000       0.000000           0.000000   
25%           0.000000         0.000000       0.000000           0.000000   
50%           0.000000         0.000000      16.000000           0.000000   
75%         950.000000       580.000000     100.000000           0.000000   
max        5060.000000      4590.000000   11570.000000        1840.000000   

           sqftdeck  ...   territorial  seattleskyline    pugetsound  \
count  18771.000000  ...  18771.000000    18771.000000  18771.000000   
mean     136.439668  ...      0.245379        0.024346      0.057536   
std      212.802391  ...      0.776513        0.260923      0.385945   
min        0.000000  ...      0.000000        0.000000      0.000000   
25%        0.000000  ...      0.000000        0.000000      0.000000   
50%        0.000000  ...      0.000000        0.000000      0.000000   
75%      210.000000  ...      0.000000        0.000000      0.000000   
max     2250.000000  ...      4.000000        4.000000      4.000000   

       lakewashington  lakesammamish  smalllakerivercreek     otherview  \
count    18771.000000   18771.000000         18771.000000  18771.000000   
mean         0.061265       0.013798             0.022215      0.014650   
std          0.407413       0.201675             0.256792      0.189631   
min          0.000000       0.000000             0.000000      0.000000   
25%          0.000000       0.000000             0.000000      0.000000   
50%          0.000000       0.000000             0.000000      0.000000   
75%          0.000000       0.000000             0.000000      0.000000   
max          4.000000       4.000000             4.000000      4.000000   

       wfntlocation  trafficnoise  airportnoise  
count  18771.000000  18771.000000       18771.0  
mean       0.106601      0.200202           0.0  
std        0.868427      0.545121           0.0  
min        0.000000      0.000000           0.0  
25%        0.000000      0.000000           0.0  
50%        0.000000      0.000000           0.0  
75%        0.000000      0.000000           0.0  
max        9.000000      3.000000           0.0  

[8 rows x 33 columns]

In [43]:
df.corr()

saleprice  sqfttotliving   yrbuilt  yrrenovated  \
saleprice             1.000000       0.621515  0.084952     0.115198   
sqfttotliving         0.621515       1.000000  0.335491     0.047570   
yrbuilt               0.084952       0.335491  1.000000    -0.254773   
yrrenovated           0.115198       0.047570 -0.254773     1.000000   
bedrooms              0.271100       0.620835  0.199428     0.012806   
sqfttotbasement       0.272623       0.312401 -0.222096     0.068928   
sqftfinbasement       0.291167       0.417700 -0.111449     0.085905   
sqftopenporch         0.276026       0.361743  0.149491     0.021661   
sqftenclosedporch     0.009184      -0.002735 -0.082004     0.019251   
sqftdeck              0.208910       0.243078  0.043985     0.057772   
heatsystem            0.135627       0.232757  0.178479    -0.002726   
heatsource           -0.052802      -0.031207  0.152118     0.039262   
bathhalfcount         0.219988       0.359931  0.387476    -0.053472   
bath3qtrcount         0.231525       0.264300  0.049133     0.047946   
bathfullcount         0.321765       0.571658  0.392164     0.034868   
condition            -0.047447      -0.104770 -0.340350    -0.090301   
sqftgarageattached    0.168607       0.458314  0.426372    -0.090000   
finbasementgrade      0.234935       0.245732 -0.035523     0.051066   
hbuasifvacant        -0.006878      -0.037365 -0.091796     0.008243   
inadequateparking     0.050667       0.023642  0.033515     0.014254   
mtrainier             0.074170       0.067232 -0.004782     0.038297   
olympics              0.152614       0.123806 -0.044274     0.052529   
cascades              0.123368       0.129707 -0.026436     0.041944   
territorial           0.312686       0.237006 -0.059308     0.098189   
seattleskyline        0.184895       0.100614 -0.020626     0.037425   
pugetsound            0.105652       0.084990 -0.067934     0.075304   
lakewashington        0.378633       0.172784 -0.044486     0.072494   
lakesammamish         0.156440       0.096768  0.015331     0.028829   
smalllakerivercreek   0.041930       0.027541 -0.008539     0.019921   
otherview             0.104786       0.081521  0.025971    -0.002297   
wfntlocation          0.261991       0.090385 -0.010418     0.051258   
trafficnoise         -0.002572      -0.057904 -0.099748     0.035666   
airportnoise               NaN            NaN       NaN          NaN   

                     bedrooms  sqfttotbasement  sqftfinbasement  \
saleprice            0.271100         0.272623         0.291167   
sqfttotliving        0.620835         0.312401         0.417700   
yrbuilt              0.199428        -0.222096        -0.111449   
yrrenovated          0.012806         0.068928         0.085905   
bedrooms             1.000000         0.223030         0.297207   
sqfttotbasement      0.223030         1.000000         0.853893   
sqftfinbasement      0.297207         0.853893         1.000000   
sqftopenporch        0.177061        -0.003292         0.017181   
sqftenclosedporch   -0.016424         0.015017         0.007393   
sqftdeck             0.100598         0.273884         0.296660   
heatsystem           0.174261         0.100265         0.106260   
heatsource          -0.050755        -0.097690        -0.019055   
bathhalfcount        0.179039        -0.051622        -0.017488   
bath3qtrcount        0.213785         0.285500         0.313235   
bathfullcount        0.444476         0.060403         0.131022   
condition            0.003031         0.095096         0.136330   
sqftgarageattached   0.283551        -0.327704        -0.163103   
finbasementgrade     0.164766         0.805154         0.827203   
hbuasifvacant       -0.008240         0.021858        -0.002950   
inadequateparking   -0.000003        -0.006937         0.000664   
mtrainier            0.015344         0.041910         0.052252   
olympics             0.024387         0.161390         0.161668   
cascades             0.043038  

In [30]:
df.to_csv('../data/our_dataframe.csv')

In [31]:
prop_res.to_csv('../data/df_with_all_columns.csv')